File: tools/lib/generation.py
from __future__ import annotations

from datetime import date
from typing import Any

from . import llm as llm_mod
from .template_data import load_outline


def _work_item_prompt_block(work_item: dict[str, Any], extra: str | None) -> str:
    lines = [
        f"Jira key: {work_item.get('key')}",
        f"Summary: {work_item.get('summary')}",
        f"Description:\n{work_item.get('description') or '_(empty)_'}",
        f"Labels: {', '.join(work_item.get('labels') or [])}",
        f"Type: {work_item.get('issue_type')}",
        f"Status: {work_item.get('status')}",
        f"Priority: {work_item.get('priority')}",
    ]
    if extra:
        lines.append(f"Additional context from user:\n{extra}")
    return "\n".join(lines)


def generate_test_plan(
    outline_path,
    work_item: dict[str, Any],
    llm_base_url: str,
    llm_api_key: str,
    llm_model: str,
    additional_context: str | None = None,
    prepared_by: str | None = None,
) -> dict[str, Any]:
    outline = load_outline(outline_path)
    ticket = work_item.get("key") or "TICKET"
    meta_title = work_item.get("summary") or "Test Plan"
    doc: dict[str, Any] = {
        "meta": {
            "ticket_key": ticket,
            "title": meta_title,
            "prepared_by": prepared_by or "QA Architecture Team",
            "date": str(date.today()),
            "version": "1.0",
        },
        "sections": [],
    }
    wi_block = _work_item_prompt_block(work_item, additional_context)
    for sec in sorted(outline.get("sections") or [], key=lambda s: s.get("order", 0)):
        sid = sec["id"]
        title = sec["title"]
        guidance = sec.get("llm_guidance") or ""
        user_prompt = f"""Section: {title}
Section id: {sid}

Guidance: {guidance}

Ticket data:
{wi_block}

Write the section body in Markdown. Start with no H1; you may use ### subheadings if useful."""
        body = llm_mod.chat_completion(llm_base_url, llm_api_key, llm_model, user_prompt)
        doc["sections"].append({"id": sid, "title": title, "body_markdown": body})
    return doc

File: web/lib/generation.ts
import { chatCompletion } from "./llm";
import type { TemplateOutline } from "./template";

function workItemPromptBlock(workItem: Record<string, unknown>, extra: string | null): string {
  const lines = [
    `Jira key: ${workItem.key}`,
    `Summary: ${workItem.summary}`,
    `Description:\n${workItem.description || "_(empty)_"}`,
    `Labels: ${((workItem.labels as string[]) ?? []).join(", ")}`,
    `Type: ${workItem.issue_type}`,
    `Status: ${workItem.status}`,
    `Priority: ${workItem.priority}`,
  ];
  if (extra) lines.push(`Additional context from user:\n${extra}`);
  return lines.join("\n");
}

export async function generateTestPlan(
  outline: TemplateOutline,
  workItem: Record<string, unknown>,
  llmBaseUrl: string,
  llmApiKey: string,
  llmModel: string,
  options?: { additionalContext?: string | null; preparedBy?: string | null }
): Promise<Record<string, unknown>> {
  const ticket = (workItem.key as string) || "TICKET";
  const metaTitle = (workItem.summary as string) || "Test Plan";
  const today = new Date().toISOString().slice(0, 10);
  const sections: Array<{ id: string; title: string; body_markdown: string }> = [];

  const wiBlock = workItemPromptBlock(workItem, options?.additionalContext ?? null);

  const ordered = [...outline.sections].sort((a, b) => (a.order ?? 0) - (b.order ?? 0));
  for (const sec of ordered) {
    const userPrompt = `Section: ${sec.title}
Section id: ${sec.id}

Guidance: ${sec.llm_guidance ?? ""}

Ticket data:
${wiBlock}

Write the section body in Markdown. Start with no H1; you may use ### subheadings if useful.`;
    const body = await chatCompletion(llmBaseUrl, llmApiKey, llmModel, userPrompt);
    sections.push({ id: sec.id, title: sec.title, body_markdown: body });
  }

  return {
    meta: {
      ticket_key: ticket,
      title: metaTitle,
      prepared_by: options?.preparedBy || "QA Architecture Team",
      date: today,
      version: "1.0",
    },
    sections,
  };
}
